Add argument spec validator

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Add controller side argument spec validation.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/module_utils/common/`
`lib/ansible/module_utils/basic.py`

samdoran committed:
Start work on adding argument spec validator class

samdoran commented:
bot_skip

samdoran committed:
None is the default return for .get()

samdoran committed:
Change method name

It's doing more than just checking since it's also transforming values. Maybe there
is a better word than "validate" to indicate "check and mutate to appropriate type".

samdoran committed:
Add method to validate argument types

Could this be moved to stand alone function? It's a duplicate of the _check_argument_values()
method from AnsibleModule.

samdoran committed:
Improve docs

Use rst style documentation.

samdoran committed:
Few more docs changes

samdoran committed:
Return a list of error messages rather than raising exceptions

samdoran commented on lib/ansible/module_utils/common/validator.py:
```diff
@@ -0,0 +1,177 @@
+
+            type_checker, wanted_name = get_type_validator(type)
+
+            self.validated_params[k] = type_checker(value)
```
In order to get _all_ the failures, I'm going to have to catch the TypeErrors here and return them all. Currently, this bails on the first validation failure. Something to keep in mind if I try to make this a standalone function.

samdoran committed:
Move methods to functions and collect all errors

samdoran committed:
Add handle_elements() function

samdoran committed:
Add subelement support to validate_argument_types

samdoran committed:
Change Validator interface

Pass in arg spec and module params when calling validate() rather than during instantiation.

samdoran committed:
Update validate() based on changes to validate_argument_types()

samdoran committed:
Use variable names consistent with other code

samdoran committed:
Don't clobber error messages

samdoran committed:
Fix bug in validate_argument_values()

samdoran committed:
Just return a string rather than an exception object

samdoran committed:
Add errors form validate_argument_values()

Shrews commented on lib/ansible/module_utils/common/validator.py:
```diff
@@ -0,0 +1,75 @@
+        validator = Validator()
+        validator.validate()
+        validated_parameters = validator.validated_parameters
+        errors = validator.error_messages
```
Let's provide accessor methods for these attributes instead and consider the attributes private (maybe rename them with a leading underscore to indicate they _should_ be private). This will let us change how we store this data later on, if we choose to.

samdoran replied:
That's a good idea. I will do that.

samdoran replied:
Do you have an opinion on properties vs actual methods?

```python
v.get_error_messages()

# or

v.error_messages
```

And the properites would look like:
```python
@property
def error_messages(self):
    return self._error_messages

@property
def validated_parameters(self):
    return self._validated_parameters
```

Not doing much currently, but gives us the ability to intercept these calls and change the behavior in the future.

Shrews replied:
Use of `@property` is fine, IMO. Serves the same purpose here.

Shrews commented on lib/ansible/module_utils/common/validator.py:
```diff
@@ -0,0 +1,75 @@
+        errors = validate_argument_values(self.argument_spec, self.module_parameters)
+        self.error_messages.extend(errors)
+
+        return self.validated_parameters
```
Probably should return something else instead of this. Maybe `bool` and use `true` to indicate valid, and `false` that errors were found. Your example shows that you expect the caller to access `validated_parameters` in a different manner anyway.

samdoran replied:
> Maybe bool and use true to indicate valid, and false that errors were found.

I'm not sure `True/False` makes sense, but I'll try it and out and see how it feels.

> Your example shows that you expect the caller to access validated_parameters in a different manner anyway.

I'll remove that example until the interface solidifies a bet more.

Shrews commented on lib/ansible/module_utils/common/validator.py:
```diff
@@ -0,0 +1,75 @@
+        """
+
+        self.argument_spec = arg_spec
+        self.module_parameters = module_parameters
```
Is there an advantage to storing these? Unless we plan to use them again in some other manner, I don't see this as necessary.

samdoran replied:
I was thinking it may be useful to have a copy of what was passed in. I didn't have a specific use for them, just that the caller may find it useful.

samdoran replied:
I ended up removing these attributes. They don't seem useful.

samdoran committed:
Remove example for now

samdoran committed:
Do not store arg spec and params as attributes

samdoran committed:
Create separate methods for role validation and module validation

These will need to do different things, so having specific methods will make it easier
to use this class for module arg spec validation in the future.

samdoran committed:
Add method for setting errors

samdoran committed:
Return True/False for validation

samdoran committed:
Update docs

Shrews commented on lib/ansible/module_utils/common/validator.py:
```diff
@@ -0,0 +1,94 @@
+    def validated_parameters(self):
+        return self._validated_parameters
+
+    def add_error(self, error):
```
Since we don't intend for a user of this class to call this method, let's make it "private" by renaming with a leading underscore.

samdoran committed:
Make _add_error method private

samdoran committed:
Fix some bugs in handle_elements

- Fix error message so it doesn't return the whole spec just the value that specified elements
- Properly add validated params and error messages from elements to return values

samdoran committed:
Improve function docs

samdoran committed:
Use better variable names than k, v

samdoran committed:
Add sub argument spec validation

Still needs a lot of work.

samdoran committed:
WIP Save Point

samdoran committed:
Use one method for validation

Want module and role validation to have the same behavior

samdoran committed:
Rename file that contains Validator class

samdoran committed:
Move _set_defaults to standalone function

samdoran committed:
Import AnsibleFallbackNotFond in basic.py for backwards compatibility

samdoran committed:
Incorporate set_defaults fixes

samdoran committed:
Handle fallbacks in validation

samdoran committed:
Remove fallback exception defition

It is imported from its new location

samdoran committed:
Handle aliases

samdoran committed:
Handle no_log values in Validator class

- Track no_log values in the Validator instance
- Sanitize error messages before returning
- move supporting functinos out of basic.py to avoid circular imports

samdoran committed:
Rearrange the deck chairs

samdoran committed:
Add documentation for set_defaults

samdoran committed:
Better handling of aliases, plus tests

samdoran committed:
Fix alias warnings

samdoran committed:
Display alias warnings

It may bet better to have the caller do this rather than the validation method.

samdoran committed:
Set default values

- add unit test for basic spec
- add unit test for defaults
- updated unit tests for failures

samdoran committed:
Move _set_fallbacks to standalone function

samdoran committed:
Remove displaying

This needs to happen controller side and not here.

samdoran committed:
Don't bother scrubbing warnings

They only contain parameter names at this point, not values.

samdoran committed:
Add quotes around parameter name in elements error message

This makes it consistent with how we report other parameter names, plus it makes it
easier to identify the parameter name.

samdoran committed:
Delete alternative approache to sub spec handling

Just stick with the current behavior moved to a function rather than trying to improve
how sub specs are processed.

samdoran committed:
Quote parameter name in error message

samdoran committed:
Remove unused import

samdoran committed:
Also catch TypeError when hanlding aliases

samdoran committed:
Try to get sub options parsing working

I really don't like how this is shaping up.

samdoran committed:
Pass in parameter positionally

samdoran committed:
Modify sub parameters directly

This does add a side effect of modifying what is passed in. But it's much simpler than returning
a dictionary and trying to merge that into a sub spec key when that is not known by the caller.

I still would rather have a function dedicated to handling sub specs that accepts sub parameters
and a sub spec as arguments. What we have currently is a function that takes the top level spec
then walks into each sub spec. That seems redundant and it has more information than it needs.

samdoran committed:
Add basic tests for sub spec

samdoran committed:
Fix call to validate_argument_types in sub spec validation

samdoran committed:
Make a copy of the original paremeters to avoid modifying the original

samdoran committed:
Pass in spec and parameters on instantiation rather than when calling .validate()

samdoran committed:
Rename class to ArgumentSpecValidator

samdoran committed:
Use parameters rather than module_parameters

This is for more than modules now

samdoran committed:
Modify parameters directly in validate_ functions

This greatly simplies the calls and makes the code easier to read.

samdoran committed:
Fixup up tests

bcoca commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -96,6 +122,206 @@ def _return_datastructure_name(obj):
+        deferred_removals.append((value, new_value))
+        value = new_value
+
+    elif isinstance(value, tuple(chain(integer_types, (float, bool, NoneType)))):
```
any of these as no_log values seem overkill, none of them can be said to be very secure (maybe a large number?), but true/false/none are really not something that should be obscured.

note: from preexisting code, not introduced in this PR

bcoca commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -0,0 +1,125 @@
+        :rtype: bool
+        """
+
+        self._no_log_values.update(set_fallbacks(self.argument_spec, self._validated_parameters))
```
set_fallbacks is already called in main code as well as update to no_log, do we really need it inside validate again?
values should be set before calling validation, not inside of it.

bcoca commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -0,0 +1,125 @@
+
+        alias_warnings = []
+        try:
+            alias_results, legal_inputs = handle_aliases(self.argument_spec, self._validated_parameters, alias_warnings)
```
aliases should also be handled pre validation (we can still validate orig and alias were not both provided)

samdoran committed:
No need to set internal properties in _handle_options()

I added this in a previous PR not fully understanding why it was needed. Upon further reading
of what it is doing, it has no effect for sub options and only makes sense to operate on the top
level options.

samdoran committed:
Fix set_default to reflect new interface

Rather than using pre=True/False as a parameter, the interface was changed to a set_default
parameter. Adjust the conditional to reflect this.

It could be argued that there should be a separate function rather than a toggle that seems
to negate the purpose of this function.

samdoran committed:
Minor cleanup

samdoran committed:
Rework and rename handle_elements --> validate_elements

- Make it behave a bit more like the other validate functions in that it accepts a list of errors.
- Rename the parameters to better reflect their contents.
- Fix bug by doing isinstance check on the type_checker not the wanted type. I introduced this when
  porting this to a function.

samdoran committed:
Move validate_elements next to other validate_* functions

samdoran committed:
Use validate_argument_types function

This allows _handle_elements() to be removed from AnsibleModule.

samdoran committed:
Check the correct variable to determine if the checker is a custom callable

Update check_type_string() function signature to get tests passing. Need a better solution, though.

s-hertel commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -275,3 +580,248 @@ def get_type_validator(wanted):
+    if errors is None:
+        errors = []
+
+    if no_log_values is None:
```
This is a duplicate

samdoran replied:
Thanks!

s-hertel commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -0,0 +1,126 @@
+            legal_inputs = None
+            self._add_error(to_native(e))
+
+        for option, alias in alias_warnings:
```
It looks like the other code path checks for deprecated aliases as well. Should this one do that?

samdoran replied:
I noticed that as well. I thought about moving it but didn't get to it yet. I'll try to do that.

samdoran replied:
I hadn't done that yet since I was focused on making this work for role arg spec, and that does not handle aliases.

samdoran replied:
This is proving to be harder than I thought.

samdoran replied:
Ok, I got it.

s-hertel reviewed (commented):
This was mostly a code review for synchronicity with the other code path. Looks good to me.

samdoran committed:
Add comment explaining code in _set_defaults()

samdoran committed:
Set defaults

Add some comments explaining why this is called twice

samdoran committed:
Accept arbitrary params for future compatibility

samdoran committed:
Remove duplicate parameter checking

samdoran committed:
Set defaults before checking required arguments

This matches the current behavior in AnsibleModule.

samdoran committed:
Add checks to sub spec validation

samdoran committed:
Use parameters instead of module_parameters/params

These functions are concerned with paremeters in general, not module_parameters.

Update documentation to reflect the changed parameter name.

Changing the function signatures in this way is a breaking change. Not sure if we are ok
with that or not.

samdoran committed:
Add alias deprecation handling to handle_aliases()

This was done in the AnsibleModule method. Incorprotade it into the standalone function.

samdoran committed:
Add alias deprecation handling to validate method

samdoran committed:
Add changelog
